crates/e2e-testing/src/controller.rs

```diff
@@ -0,0 +1,41 @@
+use crate::metadata_extractor::AppMetadata;
+use anyhow::Result;
+use async_trait::async_trait;
+use std::process::Output;
+
+/// defines crate::controller::Controller trait
+/// this is to enable running same set of tests
+/// using `spin up` or `Deploying to Fermyon Cloud`
+#[async_trait]
+pub trait Controller {
+    fn name(&self) -> String;
+    fn login(&self) -> Result<()>;
+    fn template_install(&self, args: Vec<&str>) -> Result<Output>;
+    fn new_app(&self, template_name: &str, app_name: &str) -> Result<Output>;
+    fn build_app(&self, app_name: &str) -> Result<Output>;
+    fn install_plugins(&self, plugins: Vec<&str>) -> Result<Output>;
```

Comment:
Nit: The implementation for these trait methods is identical in both `cloud_controller` and `spin_controller`. Is it possible to add a default implementation for these methods that both controller implementations will share? Still new to Rust so I'm not sure if this is possible.

Reply:
I am marking the conversation as resolved, but if you have any suggestion to make it  better, pls let me know. thanks